pkg/shed/db.go

```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		} else {
+			db.logger.Tracef("deleted key %s", string(key))
+		}
 		return err
+	})
+}
+
+// Count gives a count of all the keys present in the DB.
+func (db *DB) Count(ctx context.Context) (count int, err error) {
+	db.logger.Tracef("counting all keys")
+	db.metrics.TotalCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = false
+		i := txn.NewIterator(o)
+		defer i.Close()
+		for i.Rewind(); i.Valid(); i.Next() {
+			item := i.Item()
+			k := item.KeySize()
+			if k < 1 {
+				continue
+			}
+			count++
+		}
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while counting keys in DB")
+		db.logger.Debugf("error while doing count. Error : %w", err.Error())
+		db.metrics.TotalFailCount.Inc()
 	}
-	db.metrics.DeleteCounter.Inc()
-	return nil
+	db.logger.Tracef("Total count is %d", count)
+	return count, err
+}
+
+// CountPrefix gives a count of all the keys that starts with a given key prefix.
+// a nil prefix acts like the total count of the DB
+func (db *DB) CountPrefix(prefix []byte) (count int, err error) {
+	if prefix != nil {
+		db.logger.Tracef("counting all keys with prefix %s", string(prefix))
+	}
+	db.metrics.CountPrefixCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = false
+		o.PrefetchSize = 1024
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		// if prefix is nil, it is equivalent to counting from beginning
+		for i.Seek(prefix); i.ValidForPrefix(prefix); i.Next() {
+			item := i.Item()
+			k := item.Key()
+			if prefix != nil {
+				if !bytes.HasPrefix(k, prefix) {
+					break
+				}
+			}
+			count++
+		}
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while doing count. prefix : %v,", string(prefix))
+		db.logger.Debugf("error while doing count for prefix : %v, Error : %w", string(prefix), err.Error())
+		db.metrics.CountPrefixFailCount.Inc()
+	}
+	db.logger.Tracef("Total count for prefix %s is %d", string(prefix), count)
+	return count, err
+}
+
+// CountFrom gives a count of all the keys that start from a given prefix till the end of the DB.
+func (db *DB) CountFrom(prefix []byte) (count int, err error) {
+	if prefix != nil {
+		db.logger.Tracef("counting all keys from prefix %s", string(prefix))
+	}
+
+	db.metrics.CountFromCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = false
+		o.PrefetchSize = 1024
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		// if prefix is nil, it is equivalent to counting from beginning
+		for i.Seek(prefix); i.Valid(); i.Next() {
+			count++
+		}
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error doing count from. prefix : %v", string(prefix))
+		db.logger.Debugf("error while doing count from. prefix : %v, Error : %w", string(prefix), err.Error())
+		db.metrics.CountPrefixFailCount.Inc()
```

Comment:
Should this block return and avoid the Trace log line below?

Reply:
no. I wanted to print the count if trace is enabled.

Reply:
Should the count be irrelevant if there is an error?

Reply:
add this in an else.